frame/support/procedural/src/key_prefix.rs

```diff
@@ -0,0 +1,123 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use proc_macro2::TokenStream;
+use quote::{ToTokens, format_ident, quote};
+use std::collections::BTreeSet;
+use syn::{Ident, Result, Token, parse::Parser, punctuated::Punctuated};
+
+pub fn impl_key_prefix_for(input: proc_macro::TokenStream) -> Result<TokenStream> {
+	let args = Punctuated::<Ident, Token![,]>::parse_terminated.parse(input)?;
+	let arity = args.len();
+	if arity > 18 {
+		let msg = "Only supports implementing for tuples up to 18 elements";
+		return Err(syn::Error::new(args.last().unwrap().span(), msg));
+	}
+
+	let tuple_set = args
+		.into_iter()
+		.try_fold(BTreeSet::new(), |mut acc, ident| {
+			let span = ident.span();
+			if !acc.insert(ident) {
+				return Err(syn::Error::new(span, "Duplicate identifier name"));
+			}
+			Ok(acc)
+		})?;
+	let mut prefix_set_iter = tuple_set.iter().rev().cloned().peekable();
+	let mut suffix_set = BTreeSet::new();
+
+	let mut all_trait_impls = TokenStream::new();
+
+	while let Some(next_ident) = prefix_set_iter.next() {
+		if prefix_set_iter.peek().is_none() {
+			break;
+		}
+
+		suffix_set.insert(next_ident);
+		let hashers1 = tuple_set.iter().map(|ident| format_ident!("Hasher{}", ident));
+		let hashers2 = hashers1.clone();
+		let hashers3 = hashers1.clone();
+		let hashers4 = hashers1.clone();
+		let kargs1 = prefix_set_iter.clone().map(|ident| format_ident!("KArg{}", ident));
+		let kargs2 = kargs1.clone();
+		let kargs3 = kargs1.clone();
+		let kargs4 = kargs1.clone();
+		let kargs5 = kargs1.clone();
+		let all_idents1 = tuple_set.iter();
+		let all_idents2 = all_idents1.clone();
+		let all_idents3 = all_idents1.clone();
+		let all_idents4 = all_idents1.clone();
+		let prefixes1 = prefix_set_iter.clone();
+		let prefixes2 = prefixes1.clone();
```

Comment:
I keep on thinking to myself that there's a better way of doing this instead of cloning and creating so many let bindings, but I can't seem to think of anything...

Reply:
Does `quote!` actually take ownership of its captures? If yes, then cloning might be your only option. However, I suspect that it does not. [`to_tokens`](https://docs.rs/quote/1.0.9/quote/trait.ToTokens.html#tymethod.to_tokens) only takes `&self`.

Reply:
It does for repetitions, not so for ordinary tokens, and it makes sense, since the iterator is consumed (by repeatedly calling `next()`) after referencing it once in the `quote!` macro.

Reply:
Oh, so in earlier iterations of the code, I used a `HashSet` for most of my collections, and `quote!` didn't like it when I put the `HashSet` in a repetition, but ever since I've changed into using `BTreeSet`s, it's now possible to just put a `&BTreeSet` into a repetition and it works magically.

Source: https://docs.rs/quote/1.0.9/quote/macro.quote.html#interpolation